### ISSUE TITLE
Add delegate protocol

### ISSUE THREAD

**Issue by rlaurb** [nicklockwood/LRUCache#6]
I am interested in using LRUCache<String, URL> to build a file caching scheme. But it seems to me that I need a delegate which is called whenever the LRUCache removes a Value (as part of its internal operation). I.e., in my use-case, when LRUCache removes a URL, I need to be notified so I can remove the physical file the URL points to.

I am currently running Xcode 14.0 beta and I don't understand all the Swift 5.7 changes well enough to modify LRUCache to include this delegate. Could you help??

**Comment by nicklockwood** [comment]
Hi! I have some experience with building disk caches, but normally when you have an in-memory cache backed by a disk cache, you don't want to remove the file from disk whenever it's removed from memory - instead you have separate in-memory and on-disk capacities so that if a file is evicted from the (relatively small) memory it can still be quickly retrieved from the (much larger) on-disk cache.

But perhaps I'm not quite understanding what it is you are trying to do?

**Comment by rlaurb** [comment]
Let me step back a bit.

My CoreData-based app includes photos as on part of its data. These photos are stored in the database, but (following some guidance from Apple) I want to cache recently used images in a file cache so that I can avoid bringing the Core Data object that stores the image data into memory.

However, the Apple sample code has no logic to limit the number of files in this cache nor the total occupied size of this cache, so I was hoping to rectify this situation:
Each photo is represented in the CoreData database as two objects (Photo and ImageData). Photo has a UUID and other info, while ImageData (in a one-to-one relationship) holds the image data. So I can access a particular Photo record (which gives me the UUID) without bringing the image data into memory.
Each entry in the image FILE cache corresponds to an entry in the LRUCache<UUID, URL>. If value(forKey:) returns a non-nil URL, we use the corresponding file as the image data source; else we get the image data from the Photo’s associated ImageData object. (I’m leaving out all the obvious details).
When adding a [UUID : URL] to the LRUCache, the cost is the file size, so I can set a totalCostLimit to control the maximum size occupied by the cache directory.
I have a mechanism for reconstructing the cache during app-initialization, so the LRUCache doesn’t need to be persisted.

But in this scenario, the LRUCache has the ability to remove a cache entry when the totalCostLimit is exceeded (or when it decides to do a removalAll()) without notifying me. Under these circumstances I need to know that it did so, so that I can remove the corresponding file.

In other words, instead of having an in-memory cache which is backed by a disk cache (as you mentioned), I want to use the LRUCache as a CONTROLLER for the disk cache (which backs up the CoreData store). Sort of the other way around to what you’re thinking of.

Now maybe this is a dumb idea — it wouldn’t be the first time! — but it seems to be a pretty clean approach in this specialized circumstance. A delegate (analogous to NSCacheDelegate) would provide a nice solution. My problem is that I’m getting all tangled up in the new Swift 5.7 syntax, which I obviously haven’t figured out yet.

I hope this helps you understand what I’m trying to do.

Cheers,

Rick Aurbach
“The man who does not read has no advantage over the man who cannot read.”
– Mark Twain

**Comment by nicklockwood** [comment]
Got it. OK, well I can add a delegate as you suggest, although it will take me a little while to think about the API for it.

Alternatively, I think you can probably implement this on top of the existing LRUCache API by creating an intermediate file wrapper object and deleting the referenced file in its deinit. Something like this:

```swift
class FileWrapper {
  let url: URL

  init(url: URL) {
    self.url = url
  }

  deinit {
    try? FileManager.default.removeItem(atPath: url.path) // or whatever
  }
}

let cache = LRUCache<UUID: FileWrapper>()
```

As soon as the cache evicts a given FileWrapper instance (assuming you aren't holding onto a copy somewhere) its deinit will be called and you can either delete the file directly in there (as I've done in the example above), or delegate to some other object to do it.

The FileWrapper class might also be a good place to encapsulate some of the fetching logic you mentioned above.

Hope that all makes sense.

**Comment by rlaurb** [comment]
You are absolutely right! This is a MUCH better approach. And it handles the removeAll() case very cleanly as well.

Thank you.

Cheers,

Rick Aurbach

“Pedantry and mastery are opposite attitudes toward rules. To apply a rule to the letter, rigidly, unquestioningly, in cases where it fits and in cases where it does not fit, is pedantry ... To apply a rule with natural ease, with judgment, noticing the cases where it fits, and without ever letting the words of the rule obscure the purpose of the action or the opportunities of the situation, is mastery.”
— George Polya, mathematician (13 Dec 1887-1985)

**Comment by rlaurb** [comment]
Nick,

I just added a small method to LRUCache. I’ll explain why after showing the code.

	/// Return an array of the values currently in the cache
	func allValues() -> [Value] {
		lock.lock()
		defer { lock.unlock() }
		return values.compactMap { $0.value.value }
	}

Now the rationale:
Your file wrapper will delete the contents of the on-disk cache folder on application termination. I’d prefer to persist the on-disk cache between invocations of my app. (Fortunately, I have an algorithm for reconstructing the LRUCache on startup, which makes this feasible.) But for this to work, I need to disable the “delete file on deinit” logic at app termination. So I add a boolean to control that. Then in the FileCache class that holds the LRUCache, I implement a deinit that uses this new method to set the deleteOnDeinit flag to false, and thereby preventing files from being deleted when I don’t want them to.


Cheers,

Rick Aurbach

It is a capital mistake to theorize before one has data. Insensibly one begins to twist facts to suit theories, instead of theories to suit facts.
–Arthur Conan Doyle, physician and writer (22 May 1859-1930)

**Comment by nicklockwood** [comment]
@rlaurb I've added an `allValues` property to version 1.0.4

**Comment by aehlke** [comment]
@rlaurb would be amazing if you could open source your disk cache work, thanks for considering

**Comment by aehlke** [comment]
@nicklockwood I've reviewed his code (thanks rlaurb for providing it) and your proposed solution, and it would still be great to have a delegate to be notified of eviction to be more flexible with using file backed caching for a variety of types more easily. thanks

**Comment by aehlke** [comment]
For posterity sake here is my version of a file-backed LRUCache store

```
import Foundation
import LRUCache

open class LRUFileCache<I: Hashable & CustomStringConvertible, O: Codable>: ObservableObject {
    @Published public var cacheDirectory: URL
    private let cache: LRUCache<String, O>
    
    public init(namespace: String, version: Int, totalBytesLimit: Int = .max, countLimit: Int = .max) {
        assert(!namespace.isEmpty, "LRUFileCache namespace must not be empty")
        
        let fileManager = FileManager.default
        let cacheRoot = fileManager.urls(for: .cachesDirectory, in: .userDomainMask).first!
        cacheDirectory = cacheRoot.appendingPathComponent(namespace)
        
        cache = LRUCache(totalCostLimit: totalBytesLimit, countLimit: countLimit)
        
        if !fileManager.fileExists(atPath: cacheDirectory.path) {
            try? fileManager.createDirectory(at: cacheDirectory, withIntermediateDirectories: true)
        }
        
        let versionFileURL = cacheDirectory.appendingPathComponent(".lru_cache_version")
        if let versionData = try? Data(contentsOf: versionFileURL), String(data: versionData, encoding: .utf8) != "\(version)" {
            try? fileManager.removeItem(at: cacheDirectory)
            try? fileManager.createDirectory(at: cacheDirectory, withIntermediateDirectories: true)
        }
        try? "\(version)".data(using: .utf8)?.write(to: versionFileURL)
        
        rebuild()
    }
    
    private func cacheURL(forKey key: I) -> URL {
        return cacheDirectory.appendingPathComponent(key.description).appendingPathExtension("json")
    }
    
    public func value(forKey key: I) -> O? {
        return cache.value(forKey: String(describing: key))
    }
    
    public func setValue(_ value: O?, forKey key: I) {
        let keyString = String(describing: key)
        let url = cacheURL(forKey: key)
        if let value = value {
            if let encoded = try? JSONEncoder().encode(value) {
                try? encoded.write(to: url, options: .atomic)
                cache.setValue(value, forKey: keyString, cost: encoded.count)
            }
        } else {
            try? FileManager.default.removeItem(at: url)
            cache.removeValue(forKey: keyString)
        }
    }
    
    private func rebuild() {
        DispatchQueue.global().async { [weak self] in
            guard let self = self else { return }
            let fileManager = FileManager.default
            if let contents = try? fileManager.contentsOfDirectory(at: self.cacheDirectory, includingPropertiesForKeys: nil, options: .skipsHiddenFiles) {
                for item in contents where item.pathExtension == "json" {
                    if let data = try? Data(contentsOf: item), let decodedValue = try? JSONDecoder().decode(O.self, from: data) {
                        let keyString = item.deletingPathExtension().lastPathComponent
                        self.cache.setValue(decodedValue, forKey: keyString, cost: data.count)
                    }
                }
            }
        }
    }
    
    private func deleteOrphanFiles() throws {
        let fileManager = FileManager.default
        let contents = try fileManager.contentsOfDirectory(at: cacheDirectory, includingPropertiesForKeys: nil)
        let cacheKeys = Set(cache.allKeys)
        
        for item in contents where item.pathExtension == "json" {
            let key = item.deletingPathExtension().lastPathComponent
            if !cacheKeys.contains(key) {
                try fileManager.removeItem(at: item)
            }
        }
    }
}
```
